client/normal_value.go

```diff
@@ -0,0 +1,1531 @@
+			intVal = int64(v)
+		default:
+			return nil, NewCanNotNormalizeValue(arr)
+		}
+		result[i] = immutable.Some(intVal)
+	}
+	return NewNormalNillableIntArray(result), nil
+}
+
+func convertAnyArrToFloatArr(arr []any) (NormalValue, error) {
+	result := make([]float64, len(arr))
+	for i := range arr {
+		if arr[i] == nil {
+			return convertAnyArrToNillableFloatArr(arr)
+		}
+
+		var floatVal float64
+		switch v := arr[i].(type) {
+		case float64:
+			floatVal = v
+		case float32:
+			floatVal = float64(v)
+		case int8:
+			floatVal = float64(v)
+		case int16:
+			floatVal = float64(v)
+		case int32:
+			floatVal = float64(v)
+		case int64:
+			floatVal = float64(v)
+		case int:
+			floatVal = float64(v)
+		case uint8:
+			floatVal = float64(v)
+		case uint16:
+			floatVal = float64(v)
+		case uint32:
+			floatVal = float64(v)
+		case uint64:
+			floatVal = float64(v)
+		case uint:
+			floatVal = float64(v)
+		default:
+			return nil, NewCanNotNormalizeValue(arr)
+		}
+		result[i] = floatVal
+	}
+	return NewNormalFloatArray(result), nil
+}
+
+func convertAnyArrToNillableFloatArr(arr []any) (NormalValue, error) {
+	result := make([]immutable.Option[float64], len(arr))
+	for i := range arr {
+		if arr[i] == nil {
+			result[i] = immutable.None[float64]()
+			continue
+		}
+		var floatVal float64
+		switch v := arr[i].(type) {
+		case float64:
+			floatVal = v
+		case float32:
+			floatVal = float64(v)
+		case int8:
+			floatVal = float64(v)
+		case int16:
+			floatVal = float64(v)
+		case int32:
+			floatVal = float64(v)
+		case int64:
+			floatVal = float64(v)
+		case int:
+			floatVal = float64(v)
+		case uint8:
+			floatVal = float64(v)
+		case uint16:
+			floatVal = float64(v)
+		case uint32:
+			floatVal = float64(v)
+		case uint64:
+			floatVal = float64(v)
+		case uint:
+			floatVal = float64(v)
+		default:
+			return nil, NewCanNotNormalizeValue(arr)
+		}
+		result[i] = immutable.Some(floatVal)
+	}
+	return NewNormalNillableFloatArray(result), nil
+}
+
+func convertAnyArrToTypedArr[T any](
+	arr []any,
+	newNormalArr func([]T) NormalValue,
+	newNormalNillableArr func([]immutable.Option[T]) NormalValue,
+) (NormalValue, error) {
+	result := make([]T, len(arr))
+	for i := range arr {
+		if arr[i] == nil {
+			return convertAnyArrToNillableTypedArr[T](arr, newNormalNillableArr)
+		}
+		if v, ok := arr[i].(T); ok {
+			result[i] = v
+		} else {
+			return nil, NewCanNotNormalizeValue(arr)
+		}
+	}
+	return newNormalArr(result), nil
+}
+
+func convertAnyArrToNillableTypedArr[T any](
+	arr []any,
+	newNormalNillableArr func([]immutable.Option[T]) NormalValue,
+) (NormalValue, error) {
+	result := make([]immutable.Option[T], len(arr))
+	for i := range arr {
+		if arr[i] == nil {
+			result[i] = immutable.None[T]()
+			continue
+		}
+		if v, ok := arr[i].(T); ok {
+			result[i] = immutable.Some(v)
+		} else {
+			return nil, NewCanNotNormalizeValue(arr)
+		}
+	}
+	return newNormalNillableArr(result), nil
+}
+
+func NewNormalNil() NormalValue {
```

Comment:
thought: I don't think this should exist no?  If a `nil` is required, of an unknown type, would `NormalVoid` not be more appropriate?

Reply:
NormalVoid is like a dummy returning default value like `false`. It there just to rid other type of necessary to implement all methods.
And I think having a dedicated `NewNormalNil` more explicitly conveys the intent.

Reply:
> NormalVoid is like a dummy returning default value like false.

Yes, that's what I mean, is why I'd prefer it over `NormalNil` - `NormalNil` looks like it is typed, when it is not really - it also returns true for IsNillable no? (unwanted IMO if the type is undefined).

Reply:
It seems that everywhere this is used, we have access to the field kind which means that we can get the nillable supported type directly instead of a catch all `nil` type. I think all we would need is an extra `NewNormalValueFromKind` function.

Reply:
`DecodeFieldValue` would need to take a FieldKind as parameter as a side effect.